opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java

```diff
@@ -0,0 +1,412 @@
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+
+import opennlp.tools.util.MutableInt;
+import opennlp.tools.util.normalizer.AggregateCharSequenceNormalizer;
+import opennlp.tools.util.normalizer.CharSequenceNormalizer;
+import opennlp.tools.util.normalizer.EmojiCharSequenceNormalizer;
+import opennlp.tools.util.normalizer.NumberCharSequenceNormalizer;
+import opennlp.tools.util.normalizer.ShrinkCharSequenceNormalizer;
+import opennlp.tools.util.normalizer.TwitterCharSequenceNormalizer;
+import opennlp.tools.util.normalizer.UrlCharSequenceNormalizer;
+
+/**
+ * Implements learnable Language Detector.
+ * <p>
+ * Starts at the beginning of the charsequence and runs language
+ * detection on chunks of text.  If the end of the
+ * string is reached or there are {@link #minConsecImprovements}
+ * consecutive predictions for the best language and the confidence
+ * increases over those last predictions and if the difference
+ * in confidence between the highest confidence language
+ * and the second highest confidence language is greater than {@link #minDiff},
+ * the language detector will stop and report the results.
+ * </p>
+ * <p>
+ * The authors wish to thank Ken Krugler and
+ * <a href="https://github.com/kkrugler/yalder">Yalder</a>}
+ * for the inspiration for many of the design
+ * components of this detector.
+ * </p>
+ *
+ */
+public class ProbingLanguageDetectorME extends LanguageDetectorME {
+
+  /**
+   * Default chunk size (in codepoints) to take from the
+   * initial String
+   */
+  public static final int DEFAULT_CHUNK_SIZE = 300;
+
+  /**
+   * Default minimum consecutive improvements in confidence.
+   * If the best language is the same over this many consecutive
+   * probes, and if the confidence did not go down over those probes,
+   * the detector stops early.
+   */
+  public static final int DEFAULT_MIN_CONSEC_IMPROVEMENTS = 2;
+
+  /**
+   * Default minimum difference in confidence between the language with
+   * the highest confidence and the language with the second highest confidence.
+   */
+  public static final double DEFAULT_MIN_DIFF = 0.10;
+
+  /**
+   * Default absolute maximum length of the String (in codepoints) to process
+   */
+  public static final int DEFAULT_MAX_LENGTH = 10000;
+
+  private static final String SPACE = " ";
+
+  //size at which to break strings for detection (in codepoints)
+  private int chunkSize = DEFAULT_CHUNK_SIZE;
+
+  //require that the "best" language be the same
+  //and that the confidence in that language increase over
+  //this number of probes.
+  private int minConsecImprovements = DEFAULT_MIN_CONSEC_IMPROVEMENTS;
+
+  //Minimum difference in confidence between the best candidate
+  //and the second best candidate
+  private double minDiff = DEFAULT_MIN_DIFF;
+
+  /**
+   * Absolute maximum length (in codepoints) that will processed
+   */
+  private int maxLength = DEFAULT_MAX_LENGTH;
+
+  private CharSequenceNormalizer normalizer;
```

Comment:
The normalizer is already part of the LanguageDetectorME and shouldn't be created here, also it shouldn't be possible to modify it. The normalizer is declared at training time, and also can only be changed at training time. The reason for that is that the normalizer has an influence on the features and these have to match the ones at training time.

Reply:
Y.  Got it.